File: backend/app/core/genConsts.py
"""
Centralized General Constants (GenConsts) - KitePulse AI Backend.
Eliminates hardcoded strings, magic numbers, and duplicate values across the application.
"""

from typing import Final


class GenConsts:
    """Central constants namespace for KitePulse AI backend."""

    # Application Metadata
    APP_NAME: Final[str] = "KitePulse AI"
    APP_VERSION: Final[str] = "1.0.0"

    # Filesystem & Database
    DB_NAME: Final[str] = "trading.db"
    INSTRUMENTS_FILE_NAME: Final[str] = "instruments.json"

    # API Routes & Headers
    API_V1_PREFIX: Final[str] = "/api"
    WS_LIVE_PATH: Final[str] = "/ws/live"
    AUTH_HEADER: Final[str] = "Authorization"
    BEARER_PREFIX: Final[str] = "Bearer "

    # Trading Execution Parameters
    DEFAULT_LOT_SIZE: Final[int] = 1
    MAX_LOT_SIZE: Final[int] = 1
    DEFAULT_PRODUCT: Final[str] = "MIS"  # Intraday
    PRODUCT_CNC: Final[str] = "CNC"      # Delivery
    INITIAL_PAPER_CAPITAL: Final[float] = 100000.0

    # Trading Modes
    MODE_PAPER: Final[str] = "PAPER"
    MODE_LIVE: Final[str] = "LIVE"

    # Instrument Classifications
    TYPE_INDEX: Final[str] = "INDEX"
    TYPE_EQUITY: Final[str] = "EQUITY"

    # Order & Position Statuses
    STATUS_OPEN: Final[str] = "OPEN"
    STATUS_CLOSED: Final[str] = "CLOSED"
    STATUS_PENDING_APPROVAL: Final[str] = "PENDING_APPROVAL"
    STATUS_REJECTED: Final[str] = "REJECTED"
    STATUS_EXECUTED: Final[str] = "EXECUTED"
    STATUS_CANCELLED: Final[str] = "CANCELLED"

    # Trade Signal Types
    SIGNAL_BUY_CALL: Final[str] = "BUY_CALL"
    SIGNAL_BUY_PUT: Final[str] = "BUY_PUT"
    SIGNAL_BUY_EQUITY: Final[str] = "BUY_EQUITY"
    SIGNAL_SHORT_EQUITY: Final[str] = "SHORT_EQUITY"

    # Market Directions
    DIRECTION_BULLISH: Final[str] = "BULLISH"
    DIRECTION_BEARISH: Final[str] = "BEARISH"

    # WebSocket Event Types
    WS_EVENT_TICK: Final[str] = "TICK"
    WS_EVENT_INIT: Final[str] = "INIT"
    WS_EVENT_PROPOSAL_ALERT: Final[str] = "TRADE_PROPOSAL_ALERT"
    WS_EVENT_ORDER_UPDATE: Final[str] = "ORDER_UPDATE"
    WS_EVENT_POSITION_UPDATE: Final[str] = "POSITION_UPDATE"
    WS_EVENT_PING: Final[str] = "PING"
    WS_EVENT_PONG: Final[str] = "PONG"

    # Intervals & Timeouts (Seconds)
    KITE_TOKEN_CHECK_INTERVAL_SEC: Final[int] = 30
    MARKET_TICK_STREAM_INTERVAL_SEC: Final[int] = 2
    TOKEN_EXPIRY_HOURS: Final[int] = 24
    DB_LOCK_TIMEOUT_SEC: Final[int] = 10

    # Market Data & Chart Statuses
    CHART_STATUS_OK: Final[str] = "OK"
    CHART_STATUS_NO_DATA: Final[str] = "NO_DATA"
    CHART_STATUS_CONFIG_REQUIRED: Final[str] = "CONFIG_REQUIRED"
    CHART_STATUS_BROKER_ERROR: Final[str] = "KITE_BROKER_ERROR"

File: frontend/src/constants/genConsts.ts
/**
 * Centralized General Constants (GenConsts) - KitePulse AI Frontend.
 * Eliminates magic strings, hardcoded numbers, and scattered configuration keys.
 */
export const GenConsts = {
    // Session & LocalStorage Keys
    STORAGE_KEYS: {
        AUTH_TOKEN: "kitepulse_auth_token",
        OFFICE_MODE: "kitepulse_office_mode",
    },

    // Authentication States
    AUTH_STATES: {
        CHECKING: "CHECKING",
        UNINITIALIZED: "UNINITIALIZED",
        UNAUTHENTICATED: "UNAUTHENTICATED",
        AUTHENTICATED: "AUTHENTICATED",
    },

    // WebSocket Events
    WS_EVENTS: {
        TICK: "TICK",
        INIT: "INIT",
        TRADE_PROPOSAL_ALERT: "TRADE_PROPOSAL_ALERT",
        ORDER_UPDATE: "ORDER_UPDATE",
        POSITION_UPDATE: "POSITION_UPDATE",
        PING: "PING",
        PONG: "PONG",
    },

    // Timeouts and Polling Intervals (Milliseconds)
    INTERVALS: {
        KITE_VERIFY_MS: 30000,
        CANDLE_POLL_MS: 4000,
        NOTIFICATION_AUTO_CLOSE_MS: 5000,
        TOOLTIP_DELAY_MS: 200,
    },

    // Market Data & Chart Statuses
    CHART_STATUSES: {
        OK: "OK",
        NO_DATA: "NO_DATA",
        CONFIG_REQUIRED: "CONFIG_REQUIRED",
        KITE_BROKER_ERROR: "KITE_BROKER_ERROR",
    },

    // Trading Modes
    MODES: {
        PAPER: "PAPER",
        LIVE: "LIVE",
    },

    // Order & Position Statuses
    STATUSES: {
        OPEN: "OPEN",
        CLOSED: "CLOSED",
        PENDING_APPROVAL: "PENDING_APPROVAL",
        REJECTED: "REJECTED",
        EXECUTED: "EXECUTED",
        CANCELLED: "CANCELLED",
    },

    // Trade Signals
    SIGNALS: {
        BUY_CALL: "BUY_CALL",
        BUY_PUT: "BUY_PUT",
        BUY_EQUITY: "BUY_EQUITY",
        SHORT_EQUITY: "SHORT_EQUITY",
    },

    // Market Direction Bias
    DIRECTIONS: {
        BULLISH: "BULLISH",
        BEARISH: "BEARISH",
    },

    // Product Types
    PRODUCTS: {
        MIS: "MIS",
        CNC: "CNC",
    },

    // API Routes
    API_ROUTES: {
        AUTH_STATUS: "/api/auth/status",
        AUTH_SETUP: "/api/auth/setup",
        AUTH_LOGIN: "/api/auth/login",
        AUTH_LOGOUT: "/api/auth/logout",
        AUTH_VERIFY: "/api/auth/verify",
        STATUS: "/api/status",
        MARKET_QUOTES: "/api/market/quotes",
        MARKET_CANDLES: "/api/market/candles",
        AI_SCAN: "/api/ai/scan",
        TRADES_APPROVE: "/api/trades/approve",
        TRADES_REJECT: "/api/trades/reject",
        TRADES_POSITIONS: "/api/trades/positions",
        TRADES_ORDERS: "/api/trades/orders",
        TRADES_HISTORY: "/api/trades/history",
        TRADES_CLOSE_POSITION: "/api/trades/close-position",
        SETTINGS: "/api/settings",
        SETTINGS_VAULT: "/api/settings/vault",
        SETTINGS_TRADING_MODE: "/api/settings/trading-mode",
        SETTINGS_TEST_NOTIFICATIONS: "/api/settings/notifications/test",
    },
} as const;

export type AuthStateType = typeof GenConsts.AUTH_STATES[keyof typeof GenConsts.AUTH_STATES];
export type ChartStatusType = typeof GenConsts.CHART_STATUSES[keyof typeof GenConsts.CHART_STATUSES];
export type TradingModeType = typeof GenConsts.MODES[keyof typeof GenConsts.MODES];
export type OrderStatusType = typeof GenConsts.STATUSES[keyof typeof GenConsts.STATUSES];
export type SignalType = typeof GenConsts.SIGNALS[keyof typeof GenConsts.SIGNALS];
export type MarketDirectionType = typeof GenConsts.DIRECTIONS[keyof typeof GenConsts.DIRECTIONS];
